pgjdbc/src/main/java/org/postgresql/core/ParameterContext.java

```diff
@@ -0,0 +1,193 @@
+    }
+    return placeholderNames.get(i);
+  }
+
+  /**
+   * @param i 0-indexed position in the order of appearance
+   * @return The position of the placeholder in the SQL text for this placeholder index
+   */
+  public int getPlaceholderPosition(int i) {
+    if (placeholderPositions == null) {
+      throw new IllegalStateException("Call hasParameters() first.");
+    }
+    return placeholderPositions.get(i);
+  }
+
+  /**
+   * @param i 0-indexed position in the order of appearance
+   * @return The position of the placeholder in the order of first appearance of each placeholder
+   */
+  public int getPlaceholderAtPosition(int i) {
+    if (placeholderAtPosition == null) {
+      throw new IllegalStateException("Call hasParameters() first.");
+    }
+    return placeholderAtPosition.get(i);
+  }
+
+  public int getLastPlaceholderPosition() {
+    if (placeholderPositions == null) {
+      throw new IllegalStateException("Call hasParameters() first.");
+    }
+    return placeholderPositions.get(placeholderPositions.size() - 1);
+  }
+
+  /**
+   * Adds a named parameter to this ParameterContext. Once a named Parameter have been added all
+   * subsequent parameters must be named. Using named parameters enable reuse of the same parameters
+   * in several locations of the SQL text. The parameters only have to be sent to the PostgreSQL
+   * backend once per name specified. The values will be sent in the order of the first appearance
+   * of their placeholder.
+   *
+   * @param position in the SQL text where the parser captured the placeholder.
+   * @param bindName is the placeholder name captured by the parser.
+   * @return 1-indexed position in the order of first appearance of named parameters
+   * @throws SQLException if positional and named parameters are mixed.
+   */
+  public int addNamedParameter(int position, String bindName) throws SQLException {
+    checkAndSetBindStyle(BindStyle.NAMED);
+
+    if (placeholderPositions == null) {
+      placeholderPositions = new ArrayList<>();
+    }
+    placeholderPositions.add(position);
+
+    if (placeholderNames == null) {
+      placeholderNames = new ArrayList<>();
+    }
+    int bindIndex = placeholderNames.indexOf(bindName);
+    if (bindIndex == -1) {
+      bindIndex = placeholderNames.size();
+      placeholderNames.add(bindName);
+    }
+
+    if (placeholderAtPosition == null) {
+      placeholderAtPosition = new ArrayList<>();
+    }
+    placeholderAtPosition.add(bindIndex);
+    return bindIndex + 1;
+  }
+
+  /**
+   * @return Returns the number of placeholder appearances in the SQL text
+   */
+  public int placeholderCount() {
+    return placeholderPositions == null ? 0 : placeholderPositions.size();
+  }
+
+  /**
+   * @return Returns the number of parameter to be sent to the backend.
+   */
+  public int nativeParameterCount() {
+    if (placeholderNames != null) {
+      return placeholderNames.size();
+    }
+    return placeholderCount();
```

Comment:
could be ternary (like done above):
`return placeholderNames != null ? placeholderNames.size() : placeholderCount();`